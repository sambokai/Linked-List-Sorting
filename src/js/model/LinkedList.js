class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor(value) {
    this.length = 0;
    this.head = null;

    if (value) { this.append(value); }
  }

  append(value) {
    const node = new Node(value);
    let currentNode = this.head;

    if (!currentNode) {
      this.head = node;
      this.length += 1;

      return this;
    }

    while (currentNode.next) {
      currentNode = currentNode.next;
    }

    currentNode.next = node;

    this.length += 1;

    return this;
  }

  prepend(value) {
    const node = new Node(value);
    node.next = this.head;
    this.head = node;
    this.length += 1;
  }

  get(index) {
    let current = this.head;
    let count = 0;

    if (this.length === 0) {
      throw new Error('Empty list');
    }

    if (index < 0 || index > this.length) {
      throw new Error('Invalid index');
    }

    while (count < index) {
      current = current.next;
      count += 1;
    }

    return current;
  }

  remove(index) {
    let current = this.head;

    if (index < 0 || index > this.length) {
      throw new Error('Invalid index');
    }

    if (index === 0) {
      this.head = current.next;
      this.length -= 1;
      return this;
    }

    let count = 0;
    let previousNode;
    let nodeToBeDeleted;

    while (count < index) {
      previousNode = current;
      nodeToBeDeleted = current.next;
      current = current.next;
      count += 1;
    }

    previousNode.next = nodeToBeDeleted.next;
    this.length -= 1;
    return this;
  }

  getIndex(predicate) {
    let current = this.head;
    let count = 0;

    if (this.length === 0) {
      throw new Error('Empty list');
    }

    while (count < this.length) {
      if (predicate(current.value)) return count
      current = current.next;
      count += 1;
    }

    return -1;
  }

  clear() {
    this.head = null;
    this.length = 0;
  }

  filter(predicate) {
    const filtered = new LinkedList();
    // eslint-disable-next-line no-restricted-syntax
    for (const element of [...this].filter(predicate)) {
      filtered.append(element);
    }
    return filtered;
  }

  insertionSort(compare = (a, b) => a > b) {
    const items = [...this];
    for (let i = 0; i < items.length; i += 1) {
      const value = items[i];
      let j;
      for (j = i - 1; j > -1 && compare(items[j], value); j -= 1) {
        items[j + 1] = items[j];
      }
      items[j + 1] = value;
    }
    const sorted = new LinkedList();
    items.forEach(element => sorted.append(element));
    return sorted;
  }

  mergeSort(compare = (a, b) => a > b) {
    const sortedArray = LinkedList.mergeArray([...this], compare);
    const sortedList = new LinkedList();
    sortedArray.forEach(element => sortedList.append(element));
    return sortedList;
  }

  static mergeArray(arr, compare) {
    if (arr.length === 1) {
      return arr;
    }

    const middle = Math.floor(arr.length / 2);
    const left = arr.slice(0, middle);
    const right = arr.slice(middle);

    return this.merge(
      this.mergeArray(left, compare),
      this.mergeArray(right, compare),
      compare,
    );
  }

  static merge(left, right, compare) {
    const result = [];
    let indexLeft = 0;
    let indexRight = 0;

    while (indexLeft < left.length && indexRight < right.length) {
      if (compare(right[indexRight], left[indexLeft])) {
        result.push(left[indexLeft]);
        indexLeft += 1;
      } else {
        result.push(right[indexRight]);
        indexRight += 1;
      }
    }

    return result.concat(left.slice(indexLeft)).concat(right.slice(indexRight));
  }

  * [Symbol.iterator]() {
    let element = this.head;

    while (element) {
      yield element.value;
      element = element.next;
    }
  }
}

export default LinkedList;
