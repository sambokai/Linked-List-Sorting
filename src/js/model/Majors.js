const Majors = Object.freeze({
  COMPUTER_SCIENCE: 'Computer Science',
  MATHEMATICS: 'Mathematics',
  HISTORY: 'History',
  ECONOMICS: 'Economics',
});

export default Majors;
