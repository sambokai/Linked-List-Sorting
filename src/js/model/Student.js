import Majors from "./Majors";

let studentIdCounter = 0;
const incrementId = () => {
  studentIdCounter += 1;
  return studentIdCounter;
};

export const resetId = () => {studentIdCounter = 0;};


class Student {
  constructor(firstName, lastName, universityMajor) {
    if (Object.values(Majors).includes(universityMajor)) {
      this.universityMajor = universityMajor;
    } else {
      throw 'Invalid Major';
    }
    this.firstName = firstName;
    this.lastName = lastName;
    this.id = incrementId();
  }
}

export default Student;
