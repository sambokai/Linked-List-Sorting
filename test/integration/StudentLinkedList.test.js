import Student from "../../src/js/model/Student";
import LinkedList from "../../src/js/model/LinkedList";
import Majors from "../../src/js/model/Majors";

const exampleStudents = () => new LinkedList()
  .append(new Student("Sam", "Bokai", "cs"))
  .append(new Student("Yanis", "Varoufakis", "ec"))
  .append(new Student("Slavoj", "Zizek", "hi"))
  .append(new Student("George", "Berg", "cs"))
  .append(new Student("Sam", "Friday", "hi"))
  .append(new Student("Thomas", "Berg", "ma"));


describe('A LinkedList of Students', () => {
  it('should allow searching for students by specific parameters', () => {
    const students = exampleStudents();

    const search = {
      csStudents: (student) => student.universityMajor === Majors.COMPUTER_SCIENCE,
      firstnameSam: (student) => student.firstName === 'Sam',
    };

    const csStudents = students.filter(search.csStudents);
    const firstNameSamStudents= students.filter(search.firstnameSam);

    expect([...csStudents]).toMatchSnapshot("csStudents");
    expect([...firstNameSamStudents]).toMatchSnapshot("firstNameSamStudents");
  });

  it('should allow sorting students by specific properties using insertionSort', () => {
    const students = exampleStudents();

    const sortBy = {
      id: (a, b) => a.id > b.id,
      majorAlphabetically: (a, b) => a.universityMajor > b.universityMajor,
    };

    const sortedById = students.insertionSort(sortBy.id);
    const sortedByMajor = students.insertionSort(sortBy.majorAlphabetically);

    expect([...sortedById]).toMatchSnapshot("sortedById");
    expect([...sortedByMajor]).toMatchSnapshot("sortedByMajor");
  });

  it('should allow sorting students by specific properties using mergeSort', () => {
    const students = exampleStudents();

    const sortBy = {
      id: (a, b) => a.id > b.id,
      majorAlphabetically: (a, b) => a.universityMajor > b.universityMajor,
    };

    const sortedById = students.mergeSort(sortBy.id);
    const sortedByMajor = students.mergeSort(sortBy.majorAlphabetically);

    expect([...sortedById]).toMatchSnapshot("sortedById");
    expect([...sortedByMajor]).toMatchSnapshot("sortedByMajor");
  });
});
