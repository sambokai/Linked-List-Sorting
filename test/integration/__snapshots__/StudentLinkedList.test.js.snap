// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A LinkedList of Students should allow searching for students by specific parameters: csStudents 1`] = `
Array [
  Student {
    "firstName": "Sam",
    "id": 1,
    "lastName": "Bokai",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "George",
    "id": 4,
    "lastName": "Berg",
    "universityMajor": "Computer Science",
  },
]
`;

exports[`A LinkedList of Students should allow searching for students by specific parameters: firstNameSamStudents 1`] = `
Array [
  Student {
    "firstName": "Sam",
    "id": 1,
    "lastName": "Bokai",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Sam",
    "id": 5,
    "lastName": "Friday",
    "universityMajor": "History",
  },
]
`;

exports[`A LinkedList of Students should allow sorting students by specific properties using insertionSort: sortedById 1`] = `
Array [
  Student {
    "firstName": "Sam",
    "id": 1,
    "lastName": "Bokai",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Yanis",
    "id": 2,
    "lastName": "Varoufakis",
    "universityMajor": "Economics",
  },
  Student {
    "firstName": "Slavoj",
    "id": 3,
    "lastName": "Zizek",
    "universityMajor": "History",
  },
  Student {
    "firstName": "George",
    "id": 4,
    "lastName": "Berg",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Sam",
    "id": 5,
    "lastName": "Friday",
    "universityMajor": "History",
  },
  Student {
    "firstName": "Thomas",
    "id": 6,
    "lastName": "Berg",
    "universityMajor": "Mathematics",
  },
]
`;

exports[`A LinkedList of Students should allow sorting students by specific properties using insertionSort: sortedByMajor 1`] = `
Array [
  Student {
    "firstName": "Sam",
    "id": 1,
    "lastName": "Bokai",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "George",
    "id": 4,
    "lastName": "Berg",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Yanis",
    "id": 2,
    "lastName": "Varoufakis",
    "universityMajor": "Economics",
  },
  Student {
    "firstName": "Slavoj",
    "id": 3,
    "lastName": "Zizek",
    "universityMajor": "History",
  },
  Student {
    "firstName": "Sam",
    "id": 5,
    "lastName": "Friday",
    "universityMajor": "History",
  },
  Student {
    "firstName": "Thomas",
    "id": 6,
    "lastName": "Berg",
    "universityMajor": "Mathematics",
  },
]
`;

exports[`A LinkedList of Students should allow sorting students by specific properties using mergeSort: sortedById 1`] = `
Array [
  Student {
    "firstName": "Sam",
    "id": 1,
    "lastName": "Bokai",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Yanis",
    "id": 2,
    "lastName": "Varoufakis",
    "universityMajor": "Economics",
  },
  Student {
    "firstName": "Slavoj",
    "id": 3,
    "lastName": "Zizek",
    "universityMajor": "History",
  },
  Student {
    "firstName": "George",
    "id": 4,
    "lastName": "Berg",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Sam",
    "id": 5,
    "lastName": "Friday",
    "universityMajor": "History",
  },
  Student {
    "firstName": "Thomas",
    "id": 6,
    "lastName": "Berg",
    "universityMajor": "Mathematics",
  },
]
`;

exports[`A LinkedList of Students should allow sorting students by specific properties using mergeSort: sortedByMajor 1`] = `
Array [
  Student {
    "firstName": "George",
    "id": 4,
    "lastName": "Berg",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Sam",
    "id": 1,
    "lastName": "Bokai",
    "universityMajor": "Computer Science",
  },
  Student {
    "firstName": "Yanis",
    "id": 2,
    "lastName": "Varoufakis",
    "universityMajor": "Economics",
  },
  Student {
    "firstName": "Sam",
    "id": 5,
    "lastName": "Friday",
    "universityMajor": "History",
  },
  Student {
    "firstName": "Slavoj",
    "id": 3,
    "lastName": "Zizek",
    "universityMajor": "History",
  },
  Student {
    "firstName": "Thomas",
    "id": 6,
    "lastName": "Berg",
    "universityMajor": "Mathematics",
  },
]
`;
